### ISSUE TITLE
Persistent block devices not available on AppVM start during boot

### ISSUE THREAD

**Issue by grote** [QubesOS/qubes-issues#4975]
**Qubes OS version**
R4.0.1

**Affected component(s) or functionality**
Starting App VMs on boot when they have persistent block devices attachedd.

**Brief summary**
Starting App VMs on boot fails when they have persistent block devices attached as these seem not to be available when VMs get started.

**To Reproduce**
1. `qvm-create qube-test`
1. `qvm-block attach --persistent qube-test dom0:md126`
1. `qvm-prefs --set autostart True`
1. Reboot
1. Observe `qube-test` failing to start:

```
dom0 systemd[1]: Starting Start Qubes VM qube-test...
dom0 qvm-start[2949]: block device dom0:md126 not available
dom0 systemd[1]: qubes-vm@qube-test.service: Main process exited, code=exited, status=1/FAILURE
dom0 systemd[1]: Failed to start Start Qubes VM qube-test.
dom0 systemd[1]: qubes-vm@qube-test.service: Unit entered failed state.
dom0 systemd[1]: qubes-vm@qube-test.service: Failed with result 'exit-code'.
```

**Expected behavior**
The `qube-test` to start.

**Actual behavior**
The `qube-test` fails to start.

**Comment by pefu** [comment]
I'm struggling with a very similar problem on my laptop running with QubesOS 4.0.4 and all latest updates installed.
I've an additional harddisk containing a LUKS encrypted volume. I've added it to `/etc/crypttab` which worked fine.
However my _qube_ which I named `fileserver` has to be started manually again either using the command `systemctl start qubes-vm@fileserver` or using the Qube Manager menu. I tried to fix this using a new systemd unit in dom0 as a workaround but had no success.
A very nice and helpful guy from the community (Sven Schiffner) pointed me to this issue here and suggested to have a look into the class `BlockDevice` from the module `qubes-core-admin/qubes/etc/block.py` but so far I had not luck in identifying the piece of code which emits in my case the following error message to systemd:
  ` qvm-start[3361]: block device dom0:dm-6 not available`
I'm welcoming any suggestions how to proceed here.

**Comment by pefu** [comment]
I finally found out that my error message is generated in `qubes-core-admin/qubes/vm/qubesvm.py` line 1139 due to the fact that during boot for some reason still not understood by me the block device is an instance of the class `qubes.devices.UnknownDevice`. Further investigation showd me, that `self.devices` is initialized in the method `__init__()` of class `BaseVM` in line 225 of `qubes-core-admin/qubes/vm/__init__.py`. I suspect that during boot the `DeviceManager()` instance is initialized just before **dom0** is able to provide the device. However I'm still not familar enough with the code. Any hints  how to proceed welcome.

**Comment by noptys** [comment]
If the persistent block device has a device-mapper entry (`dmsetup ls`), it seems to be always be stored in qubesd with that designation, e.g. `dom0:dm-6`.  On reboot, these mappings often are different.  I'm having a similar problem with a persistent LV for sys-wyng.  Seems like the fix is likely #2272?
